src/Simplex/Chat/Library/Commands.hs

```diff
@@ -2928,53 +2946,47 @@ processChatCommand' vr = \case
+    connectViaContact :: User -> Maybe PreparedChatEntity -> IncognitoEnabled -> CreatedLinkContact -> Maybe SharedMsgId -> Maybe (SharedMsgId, MsgContent) -> CM ConnectViaContactResult
+    connectViaContact user@User {userId} preparedEntity_ incognito (CCLink cReq@(CRContactUri crData@ConnReqUriData {crClientData}) sLnk) welcomeSharedMsgId msg_ = withInvitationLock "connectViaContact" (strEncode cReq) $ do
       let groupLinkId = crClientData >>= decodeJSON >>= \(CRDataGroup gli) -> Just gli
           cReqHash = ConnReqUriHash . C.sha256Hash . strEncode
           cReqHash1 = cReqHash $ CRContactUri crData {crScheme = SSSimplex}
           cReqHash2 = cReqHash $ CRContactUri crData {crScheme = simplexChat}
-      case groupLinkId of
-        -- contact address
+      -- groupLinkId is Nothing for business chats
+      when (isJust msg_ && isJust groupLinkId) $ throwChatError CEConnReqMessageProhibited
+      case preparedEntity_ of
+        Just (PCEContact ct@Contact {activeConn}) -> case activeConn of
+          Nothing -> connect' Nothing cReqHash1 Nothing
+          Just conn@Connection {connStatus, xContactId} -> case connStatus of
+            ConnPrepared -> joinPreparedConn' xContactId conn False
+            _ -> pure $ CVRConnectedContact ct
+        Just (PCEGroup _gInfo GroupMember {activeConn}) -> case activeConn of
+          Nothing -> connect' groupLinkId cReqHash1 Nothing
+          Just conn@Connection {connStatus, xContactId} -> case connStatus of
+            ConnPrepared -> joinPreparedConn' xContactId conn $ isJust groupLinkId
+            _ -> connect' groupLinkId cReqHash1 xContactId -- why not "already connected" for host member?
         Nothing ->
           withFastStore' (\db -> getConnReqContactXContactId db vr user cReqHash1 cReqHash2) >>= \case
-            (Just Contact {activeConn = Just conn@Connection {connStatus = ConnPrepared}}, Just (xContactId, _)) -> do
-              incognitoProfile <- joinPreparedConn' xContactId conn
-              pure $ CRSentInvitation user (toPCC conn) incognitoProfile
-            (Just contact, _) -> pure $ CRContactAlreadyExists user contact
-            (Nothing, Just (xContactId, Just conn@Connection {connId, connStatus = ConnPrepared})) -> do
-              incognitoProfile <- joinPreparedConn' xContactId conn
-              -- TODO [short links] align Connection and PendingContactConnection so it doesn't need to be read again
-              pcc <- withStore $ \db -> getPendingContactConnection db userId connId
-              pure $ CRSentInvitation user pcc incognitoProfile
-            (Nothing, xContactId_) -> procCmd $ do
+            Right ct@Contact {activeConn} -> case groupLinkId of
+              Nothing -> case activeConn of
+                Just conn@Connection {connStatus = ConnPrepared, xContactId} -> joinPreparedConn' xContactId conn False
+                _ -> pure $ CVRConnectedContact ct
+              Just gLinkId ->
```

Comment:
this all as simplified because xContactId is now in Connection. It's not possble, for example, that xContact is present and Connection is not. Also flipping cases removed duplicate code, as the logic for connection in the absense of contact is the same with and without groupLinkId.